lib/checkother.cpp

```diff
@@ -2023,7 +2023,7 @@ void CheckOther::checkCharVariable()
                 if (!tok->variable()->isArray() && !tok->variable()->isPointer())
                     continue;
                 const Token *index = tok->next()->astOperand2();
-                if (warning && tok->variable()->isArray() && astIsSignedChar(index) && index->getValueGE(0x80, *mSettings))
+                if (warning && astIsSignedChar(index) && index->getValueLE(-1, *mSettings))
```

Comment:
I fear there will be false positives when there is not an array. I.e. this code is safe:

     void foo() {
          int buf[1000];
          int *p = buf + 512;
          signed char c = 0x80;
          p[c] = 123;
     }

Reply:
> I fear there will be false positives when there is not an array. I.e. this code is safe:
> 
> ```
>  void foo() {
>       int buf[1000];
>       int *p = buf + 512;
>       signed char c = 0x80;
>       p[c] = 123;
>  }
> ```

Good catch. This is indeed reported. The index will be negative and the memory access still valid but I would not call it "safe".

But we do not care if the access is actually valid because if you reduce the offset, it would not be reported.

If you remove `signed` you will get a `-Wchar-subscripts` Clang warning though.